app/src/test/java/org/mozilla/focus/cfr/CfrMiddlewareTest.kt

```diff
@@ -0,0 +1,122 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+package org.mozilla.focus.cfr
+
+import mozilla.components.browser.state.action.ContentAction
+import mozilla.components.browser.state.action.TabListAction
+import mozilla.components.browser.state.state.BrowserState
+import mozilla.components.browser.state.state.SecurityInfoState
+import mozilla.components.browser.state.state.TabSessionState
+import mozilla.components.browser.state.state.createTab
+import mozilla.components.browser.state.store.BrowserStore
+import mozilla.components.support.test.any
+import mozilla.components.support.test.ext.joinBlocking
+import mozilla.components.support.test.libstate.ext.waitUntilIdle
+import mozilla.components.support.test.robolectric.testContext
+import org.junit.Assert.assertEquals
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.mockito.Mock
+import org.mockito.Mockito.doReturn
+import org.mockito.Mockito.spy
+import org.mockito.MockitoAnnotations
+import org.mozilla.focus.Components
+import org.mozilla.focus.state.AppState
+import org.mozilla.focus.state.AppStore
+import org.mozilla.focus.state.Screen
+import org.mozilla.focus.utils.Settings
+import org.robolectric.RobolectricTestRunner
+
+@RunWith(RobolectricTestRunner::class)
+class CfrMiddlewareTest {
+    private lateinit var cfrMiddleware: CfrMiddleware
+    private lateinit var browserStore: BrowserStore
+    private lateinit var appStore: AppStore
+    private lateinit var settings: Settings
+
+    @Mock
+    private lateinit var components: Components
+
+    @Before
+    fun setUp() {
+        MockitoAnnotations.openMocks(this)
+        cfrMiddleware = spy(CfrMiddleware(components))
```

Comment:
If possible, I would strongly encourage not using spies here. There should be another way e.g. worst case by separating the `shouldShowCfrForTrackingProtection` call from the middleware, but looks like there's a simpler way?

The check is based on the action itself which is under test, and the component settings and state which can be set in the test as well so that `shouldShowCfrForTrackingProtection` either return true or false, as needed.